### PR TITLE
Modularize scrutineer

### PR DESCRIPTION
**Pull request by rock-yu**
## Proposed changes

Modularized scrutineer into core/elasticsearch7/jdbc/scrutineer(runner) modules so that we can:
1) extend the Stream implementations and pick and compose what we need
2) depend only on the Verifier module with in-house Stream implementations
3) maintain the existing capability to run rpm or install tarball through scrutineer(runner) module
4) existing clients who depend on "com.aconex.scrutineer:scrutineer" will continue to work 

**This addresses issue #53 **

## Checklist

- [ x] I've read the [CONTRIBUTING](https://github.com/Aconex/scrutineer/blob/master/CONTRIBUTING.md) doc
- [x ] All tests pass locally with my changes
- [ ] I've added unit tests and integration tests.
- [ ] I've added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by tallpsmith** [comment]
I'm just trying to work out why Travis hasn't tried to build this MR....

**Review by tallpsmith — CHANGES_REQUESTED** [review]
Can you have a look at the feedback here (fairly minor).

I wish there was a way for the import order to be consistent - we probably need to define what they should be for this Project to avoid 'diff noise'.

File: elasticsearch7/src/test/resources/pmd.xml [hunk 1]
```diff
@@ -0,0 +1,283 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (tallpsmith) [reply elasticsearch7/src/test/resources/pmd.xml]: I think you addressed this in the latest commit, but just check and resolve if you feel it already did.

**Comment by tallpsmith** [comment]
Last points for discussion is about version numbering and CHANGELOG.  If you can update the version SNAPSHOT to indicate a "major" (though not necessarily breaking) change?  I think the CHANGELOG should outline what was done too if you can add something there please.

File: core/src/main/java/com/aconex/scrutineer/javautil/ControlledTimeSource.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.aconex.scrutineer.javautil;
+
+public class ControlledTimeSource implements TimeSource {
```
[the remark this answers is not in the document]
Comment (rock-yu) [reply core/src/main/java/com/aconex/scrutineer/javautil/ControlledTimeSource.java]: Done

File: jdbc/src/test/java/com/aconex/scrutineer/jdbc/JdbcIdAndVersionStreamTest.java [hunk 3]
```diff
@@ -46,62 +41,62 @@ public void setup() {
 
     @Test
     public void shouldCloseDBResourcesEvenIfNoIteration() throws SQLException {
-        when(connection.createStatement()).thenReturn(statement);
-        when(statement.executeQuery(SQL)).thenReturn(resultSet);
-        when(resultSet.getMetaData()).thenReturn(metaData);
+        Mockito.when(connection.createStatement()).thenReturn(statement);
```
[the remark this answers is not in the document]
Comment (rock-yu) [reply jdbc/src/test/java/com/aconex/scrutineer/jdbc/JdbcIdAndVersionStreamTest.java]: Done

File: elasticsearch7/src/test/java/com/aconex/scrutineer/elasticsearch/IdAndVersionDataReaderTest.java [hunk 4]
```diff
@@ -44,8 +45,8 @@ public void shouldReadNextIdAndVersionObjectFromStream() throws IOException {
         when(objectInputStream.readUTF()).thenReturn(ID);
         when(objectInputStream.readLong()).thenReturn(VERSION);
         IdAndVersion idAndVersion = idAndVersionDataReader.readNext();
-        assertThat(idAndVersion.getId(), is(ID));
-        assertThat(idAndVersion.getVersion(), is(VERSION));
+        MatcherAssert.assertThat(idAndVersion.getId(), is(ID));
```
[the remark this answers is not in the document]
Comment (rock-yu) [reply elasticsearch7/src/test/java/com/aconex/scrutineer/elasticsearch/IdAndVersionDataReaderTest.java:48]: Done

File: elasticsearch7/src/test/java/com/aconex/scrutineer/elasticsearch/IdAndVersionInputStreamIteratorTest.java [hunk 5]
```diff
@@ -50,15 +51,15 @@ public void shouldGetTheNextItem() throws IOException {
         IdAndVersion idAndVersion = new StringIdAndVersion(ID, VERSION);
         when(idAndVersionDataReader.readNext()).thenReturn(idAndVersion);
         IdAndVersionInputStreamIterator idAndVersionInputStreamIterator = new IdAndVersionInputStreamIterator(idAndVersionDataReader);
-        assertThat(idAndVersionInputStreamIterator.next(), is(idAndVersion));
+        assertThat(idAndVersionInputStreamIterator.next(), CoreMatchers.is(idAndVersion));
```
[the remark this answers is not in the document]
Comment (rock-yu) [reply elasticsearch7/src/test/java/com/aconex/scrutineer/elasticsearch/IdAndVersionInputStreamIteratorTest.java:54]: Done

File: elasticsearch7/src/test/java/com/aconex/scrutineer/elasticsearch/IdAndVersionInputStreamIteratorTest.java [hunk 6]
```diff
@@ -50,15 +51,15 @@ public void shouldGetTheNextItem() throws IOException {
...
     }
 
     @Test
     public void hasNextShouldReturnFalseAtEndOfTheStream() throws IOException {
     	IdAndVersion idAndVersion = new StringIdAndVersion(ID, VERSION);
         when(idAndVersionDataReader.readNext()).thenReturn(idAndVersion).thenReturn(null);
         IdAndVersionInputStreamIterator idAndVersionInputStreamIterator = new IdAndVersionInputStreamIterator(idAndVersionDataReader);
-        assertThat(idAndVersionInputStreamIterator.next(), is(idAndVersion));
+        assertThat(idAndVersionInputStreamIterator.next(), CoreMatchers.is(idAndVersion));
```
[the remark this answers is not in the document]
Comment (rock-yu) [reply elasticsearch7/src/test/java/com/aconex/scrutineer/elasticsearch/IdAndVersionInputStreamIteratorTest.java:62]: Done

File: elasticsearch7/src/test/resources/pmd.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rock-yu) [reply elasticsearch7/src/test/resources/pmd.xml]: Done

File: jdbc/src/test/java/com/aconex/scrutineer/jdbc/JdbcIdAndVersionStreamTest.java [hunk 7]
```diff
@@ -75,7 +75,7 @@ public void shouldExecuteSQLQuery() throws SQLException {
         JdbcIdAndVersionStream jdbcIdAndVersionStream = new JdbcIdAndVersionStream(connection, SQL, idAndVersionFactory);
         jdbcIdAndVersionStream.open();
         IdAndVersionResultSetIterator iterator = (IdAndVersionResultSetIterator) jdbcIdAndVersionStream.iterator();
-        assertThat(iterator.getResultSet(), is(resultSet));
+        assertThat(iterator.getResultSet(), Is.is(resultSet));
```
Comment (tallpsmith) [inline jdbc/src/test/java/com/aconex/scrutineer/jdbc/JdbcIdAndVersionStreamTest.java:77]: this doesn't look like it's statically imported ?

File: jdbc/src/test/java/com/aconex/scrutineer/jdbc/JdbcIdAndVersionStreamTest.java [hunk 8]
```diff
@@ -99,7 +99,7 @@ public void shouldThrowExceptionIfResultSetFailsToClose() throws SQLException {
             when(statement.executeQuery(SQL)).thenReturn(resultSet);
         } catch (SQLException e) {
             e.printStackTrace();
-            fail("Unexpected exception");
+            Assert.fail("Unexpected exception");
```
Comment (tallpsmith) [inline jdbc/src/test/java/com/aconex/scrutineer/jdbc/JdbcIdAndVersionStreamTest.java:101]: no static import?

File: elasticsearch7/pom.xml [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>org.elasticsearch</groupId>
+            <artifactId>elasticsearch</artifactId>
+            <version>${elasticsearch.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.elasticsearch.client</groupId>
+            <artifactId>x-pack-transport</artifactId>
```
[the remark this answers is not in the document]
Comment (rock-yu) [reply elasticsearch7/pom.xml:35]: It was existing dependencies we added before. Still I checked, the x-pack security is free and open-sourced, https://www.elastic.co/blog/security-for-elasticsearch-is-now-free
https://www.elastic.co/blog/doubling-down-on-open

File: core/src/main/java/com/aconex/scrutineer/javautil/ControlledTimeSource.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rock-yu) [reply core/src/main/java/com/aconex/scrutineer/javautil/ControlledTimeSource.java]: Done, moved to Test

File: elasticsearch7/pom.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (tallpsmith) [reply elasticsearch7/pom.xml:35]: free does not imply "good license"....

[the remark this answers is not in the document]
Comment (tallpsmith) [reply elasticsearch7/pom.xml:35]: I'll create a seperate issue for that.

File: scrutineer/pom.xml [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+            <artifactId>verifier</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>jdbc</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+
+        <dependency>
```
[the remark this answers is not in the document]
Comment (rock-yu) [reply scrutineer/pom.xml:41]: This pom is actually no longer the top-level. This sub-module (stilled called scrutineer) represents the Command line runner, it runs the default JDBC against ES, so it requires all these dependencies.

**Comment by rock-yu** [comment]
> Last points for discussion is about version numbering and CHANGELOG. If you can update the version SNAPSHOT to indicate a "major" (though not necessarily breaking) change? I think the CHANGELOG should outline what was done too if you can add something there please.

**Pull request merged**